### PR TITLE
Allow injecting a profiler runtime into `#![no_core]` crates

### PR DESCRIPTION
**Pull request by Zalathar**
An alternative to #133300, allowing `-Cinstrument-coverage` to be used with `-Zbuild-std`.

The incompatibility between `profiler_builtins` and `#![no_core]` crates appears to have been caused by profiler_builtins depending on core, and therefore conflicting with core (or minicore).

But that's a false dependency, because the profiler doesn't contain any actual Rust code. So we can just mark the profiler itself as `#![no_core]`, and remove the incompatibility error.

---

For context, the error was originally added by #79958.

### CODE REVIEW

**Comment by Zalathar** [comment]
cc @Enselic @jieyouxu

**Comment by jieyouxu** [comment]
I can't say I'm too familiar with `profiler_builtins` injection either, but cutting off the unnecessary dependency to `core` seems like a less fragile option that only skipping when building an `.rlib`.

File: library/profiler_builtins/src/lib.rs [hunk 1]
```diff
@@ -1,11 +1,15 @@
-#![no_std]
+// tidy-alphabetical-start
+#![allow(internal_features)]
+#![feature(no_core)]
 #![feature(profiler_runtime)]
+#![feature(staged_api)]
+// tidy-alphabetical-end
+
+// Other attributes:
+#![no_core]
```
Comment (bjorn3) [inline library/profiler_builtins/src/lib.rs:9]: You can also remove the core and compiler_builtins dependencies in Cargo.toml.

**Comment by Zalathar** [comment]
r? compiler

It's possible to write a run-make test that actually runs cargo with `-Zbuild-std` and `RUSTFLAGS="-Cinstrument-coverage"`, but it's a bit fiddly and takes 20+ seconds to run (since it builds its own libraries), so I'm not sure whether it's worth adding.

**Comment by jieyouxu** [comment]
r? jieyouxu

File: compiler/rustc_metadata/src/creader.rs [hunk 2]
```diff
@@ -799,7 +799,7 @@ impl<'a, 'tcx> CrateLoader<'a, 'tcx> {
         self.inject_dependency_if(cnum, "a panic runtime", &|data| data.needs_panic_runtime());
     }
 
-    fn inject_profiler_runtime(&mut self, krate: &ast::Crate) {
+    fn inject_profiler_runtime(&mut self, _krate: &ast::Crate) {
```
Comment (jieyouxu) [inline compiler/rustc_metadata/src/creader.rs:802]: Question: if `_krate` isn't used... maybe we should remove it?

**Review by jieyouxu — APPROVED** [review]
Thanks, this seems reasonable to me, just one nit. You can r=me with the nit.

File: compiler/rustc_metadata/src/creader.rs (hunk 2 above)
Comment (Zalathar) [reply compiler/rustc_metadata/src/creader.rs:802]: Yeah, in hindsight my reasons for leaving it as-is don't make much sense, so I'll change it.

**Comment by Zalathar** [comment]
Addressed review feedback [(diff)](https://github.com/rust-lang/rust/compare/bd7aa4bcda901b7654b1552b04df667aa6a04387..6798ecaf1000c223378151023b8eb37f3464af29).

@bors r=jieyouxu

**Pull request merged**